plugins/pastefromword/filter/default.js

```diff
@@ -354,41 +343,128 @@
 					}
 
-					var backgroundColor = ascendantStyle[ 'background-color' ] || ascendantStyle[ 'BACKGROUND-COLOR' ];
+					var backgroundColor = ascendantStyle[ 'background-color' ];
 					if ( backgroundColor ) {
 						Style.setStyle( element, 'background-color', backgroundColor, true );
 					}
 
-					var styles = tools.parseCssText( element.attributes.style );
+					var styles = tools.parseCssText( element.attributes.style, true ),
+						borderStyles = styles.border ? CKEDITOR.tools.style.parse.border( styles.border ) : {};
 
-					for ( var style in styles ) {
-						var temp = styles[ style ];
-						delete styles[ style ];
-						styles[ style.toLowerCase() ] = temp;
-					}
+					for ( var i = 0; i < allowedBorderTypes.length; i++ ) {
+						var borderType = allowedBorderTypes[ i ],
+							borderValue = mergeBorderStyles( borderType, styles[ borderType ] );
 
-					// Count all border styles that occur in the table.
-					for ( var i = 0; i < borderStyles.length; i++ ) {
-						if ( styles[ borderStyles[ i ] ] ) {
-							var key = styles[ borderStyles[ i ] ];
-							tdBorders[ key ] = tdBorders[ key ] ? tdBorders[ key ] + 1 : 1;
+						if ( borderValue ) {
+							Style.setStyle( element, borderType, borderValue );
 						}
 					}
 
+					// Remove unused styles.
+					Style.setStyle( element, 'border', '' );
+					CKEDITOR.tools.array.forEach( borderModifiers, function( modifier ) {
+						Style.setStyle( element, 'border-' + modifier, '' );
+					} );
+
 					Style.createStyleStack( element, filter, editor,
 						/margin|text\-align|padding|list\-style\-type|width|height|border|white\-space|vertical\-align|background/i );
+
+					function mergeBorderStyles( borderType, borderValue ) {
+						var currentBorder = borderValue ? CKEDITOR.tools.style.parse.border( borderValue ) : {},
+							borderObj = CKEDITOR.tools.array.reduce( borderModifiers, function( current, modifier ) {
+
+								if ( !current[ modifier ] ) {
+									var style = getBorderStyle( modifier, borderType );
+
+									if ( style ) {
+										current[ modifier ] = style;
+									}
+								}
+
+								return current;
+							}, currentBorder );
+
+						// Use parent styles if required `style` property is `none`.
+						if ( borderObj.style === 'none' ) {
+							return null;
+						}
+
+						// Set default cell style properties.
+						borderObj = CKEDITOR.tools.object.merge( {
+							color: 'black',
+							width: '0'
+						}, borderObj );
+
+						// Width may be set to some string value e.g. `inherit`.
+						borderObj.width = parseFloat( borderObj.width ) ?
+							CKEDITOR.tools.convertToPx( borderObj.width ) + 'px'
+							: borderObj.width;
+
+						return borderObj.width + ' ' +
+							borderObj.style + ' ' +
+							// Replace old CSS2 `windowtext` color with `black`.
+							borderObj.color.replace( 'windowtext', 'black' );
+					}
+
+					function getBorderStyle( modifier, borderType ) {
```

Comment:
CSS parsing code should (re)use [`CKEDITOR.tools.style.parse`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_tools_style_parse.html) methods. In case something is missing there, just add new feature there.